packages/transaction-controller/src/utils/simulation.ts

```diff
@@ -354,62 +376,122 @@ function getTokenBalanceTransactions(
 
       tokenKeys.add(tokenKey);
 
-      const parameters = [request.from];
-
-      if (event.tokenStandard === SimulationTokenStandard.erc1155) {
-        parameters.push(tokenId as Hex);
-      }
+      const data = getBalanceTransactionData(
+        event.tokenStandard,
+        request.from,
+        tokenId,
+      );
 
       result.set(simulationToken, {
         from: request.from,
         to: event.contractAddress,
-        data: new Interface(event.abi).encodeFunctionData(
-          'balanceOf',
-          parameters,
-        ) as Hex,
+        data,
       });
     }
 
     return result;
   }, new Map<SimulationToken, SimulationRequestTransaction>());
 }
 
+/**
+ * Extract token IDs from a parsed event.
+ * @param event - The parsed event.
+ * @returns An array of token IDs.
+ */
+function getEventTokenIds(event: ParsedEvent): (Hex | undefined)[] {
+  if (event.tokenStandard === SimulationTokenStandard.erc721) {
+    return [event.args.tokenId as Hex];
+  }
+
+  if (
+    event.tokenStandard === SimulationTokenStandard.erc1155 &&
+    event.name === 'TransferSingle'
+  ) {
+    return [event.args.id as Hex];
+  }
+
+  if (
+    event.tokenStandard === SimulationTokenStandard.erc1155 &&
+    event.name === 'TransferBatch'
+  ) {
+    return event.args.ids as Hex[];
+  }
+
+  // ERC-20 does not have a token ID so default to undefined.
+  return [undefined];
+}
+
+/**
+ * Extract the value from a balance transaction response.
+ * @param from - The address to check the balance of.
+ * @param token - The token to check the balance of.
+ * @param response - The balance transaction response.
+ * @returns The value of the balance transaction.
+ */
+function getValueFromBalanceTransaction(
+  from: Hex,
+  token: SimulationToken,
+  response: SimulationResponseTransaction,
+): Hex {
+  const normalizedReturn = normalizeReturnValue(response.return);
+
+  if (token.standard === SimulationTokenStandard.erc721) {
+    return normalizedReturn === from ? '0x1' : '0x0';
+  }
+
+  return normalizedReturn;
+}
+
+/**
+ * Generate the balance transaction data for a token.
+ * @param tokenStandard - The token standard.
+ * @param from - The address to check the balance of.
+ * @param tokenId - The token ID to check the balance of.
+ * @returns The balance transaction data.
+ */
+function getBalanceTransactionData(
+  tokenStandard: SimulationTokenStandard,
+  from: Hex,
+  tokenId?: Hex,
+): Hex {
+  switch (tokenStandard) {
+    case SimulationTokenStandard.erc721:
+      return new Interface(abiERC721).encodeFunctionData('ownerOf', [
```

Comment:
It looks like we're returning the owner as the balance. If that is what is happening, could you explain the reasoning?

Reply:
Previously, we were using `balanceOf` on all token standards including ERC-721, where `balanceOf` returns the count of all token IDs owned by an account, meaning the `tokenBalanceChanges` would contain incorrect balances for each ERC-721 as the `previousBalance` and `newBalance` would reflect the total ID count.

In addition, we weren't actually verifying the user received a specific token ID, only that their total token ID count increased.

`ownerOf` returns the account address that owns a specific token ID which we convert to balance by setting it to `1` if it matches the user account, or `0` otherwise.